docs/decisions/0011-depending-on-multiple-event-bus-implementations.rst

```diff
@@ -0,0 +1,80 @@
+- More concretely, event-bus-kafka depends on confluent-kafka, but that package does not provide binaries compatible with Linux on Apple Silicon. This means that developers on Apple M1 laptops cannot install this package in devstack at all. So as a knock-on effect, event-bus-kafka has had to keep confluent-kafka out of its own in-tree dependencies; deployers wishing to use it have to install and version that package out-of-tree anyhow. While this may eventually be resolved, this does obviate a number of the stated benefits for development and version management for the Kafka users.
+
+We are not necessarily *satisfied* with this approach, but regard it as suitable for at least one release.
+
+Rejected Alternatives
+*********************
+
+There are several alternative approaches we have considered:
+
+- Have the implementation be part of the openedx-events package
+- Just have one implementation
+- Require all implementations to be installed out-of-tree
+- Depend on one default implementation in-tree, and require that others be installed
+
+Implement inside openedx-events
+===============================
+
+We briefly discussed the idea of including all of the implementations *in* the openedx-events package, possibly as extras (e.g. ``openedx-events[kafka]``). However, this would have complicated the development of openedx-events; it would have become impossible to e.g. pin the version of the Redis implementation in a dependent IDA while continuing to receive updates to the Kafka implementation. (The use of extras also would not have helped our dependencies issue; we'd then need to decide which extras to depend upon from IDAs' base requirements files.)
+
+Single implementation
+=====================
+
+It's very unlikely that one implementation would satisfy everyone. Any event store, queue, or message bus technology that is used is additional software that deployers must learn to install, manage, configure, and operate. The exceptions would be MySQL and Redis, which are already used in Open edX. However, neither of these was used for the first implementation. Therefore, even if Redis turns out to work very well, we'll still need the extension point.
+
+All out-of-tree
+===============
+
+As discussed, this results in more development friction (although possibly lower maintenance burden). The package also would not participate in our usual ``make upgrade`` process. This leads to either A) more operational risk if deployers leave the implementation package version unpinned in their deployment pipeline, or B) more toil if the version is pinned but has to be manually bumped in every IDA every time a new package version is released.
```

Comment:
I wonder if this is temporarily-rejected/deferred? Could we find other solutions that improve all of these issues **and** keep us from installing unused packages for people? I think it would take a bunch of work, but as we want to better support plugins and optional parts of the platform, improved tools on this front might be worth the investment.

What I am saying does not change the decision, but may just change how we discuss this rejection. My notes shouldn't replace anything you wrote, but might add to it.

Reply:
Yes, it's probably worth calling out that this is something we'd like to be able to do better.